unit-tests/py/rspy/devices.py

```diff
@@ -132,13 +133,69 @@ def query( monitor_changes = True ):
             else:
                 sn = dev.get_info( rs.camera_info.serial_number )
             device = Device( sn, dev )
+            if device.port is not None:
+                known_ports.append(device.port)
             _device_by_sn[sn] = device
             log.d( '... port {}:'.format( device.port is None and '?' or device.port ), dev )
     finally:
         log.debug_unindent()
     #
     if monitor_changes:
         _context.set_devices_changed_callback( _device_change_callback )
+    if acroname:
+        need_recover = False
+        for dev in _device_by_sn.values():
+            if dev.port is None:
+                need_recover = True
+        if need_recover:
+            log.d( 'trying do discover unknown ports' )
+            ports = acroname.ports()
+            unknown_ports = [port for port in ports if port not in known_ports]
+            print( "ports:", ports )
+            print( "known ports:", known_ports )
+            print( "unknown ports:", unknown_ports )
+            for known_port in known_ports:
+                if known_port not in ports:
+                    log.e( "ERROR MESSAGE" ) # log.f?
+            if len( unknown_ports ) == 1:
+                for device in _device_by_sn.values():
+                    if device.port is None:
+                        log.d( 'port', unknown_ports[0], 'has device', device.serial_number )
+                        device._port = unknown_ports[0]
+            else:
+                # disabling all ports
+                acroname.disable_ports( ports )
+                for retry in range( 5 ):
+                    if len( enabled() ) == 0:
+                        break
+                    time.sleep( 1 )
+                if len( enabled() ) != 0:
+                    log.w( 'Could not disable ports, can\'t infer ports for recovery devices' )
+                    return
+                # enabling one port at a time to try and find what device is connected to it
+                for port in unknown_ports:
+                    acroname.enable_ports( [port], disable_other_ports=True )
+                    sn = None
+                    for retry in range( 5 ):
+                        if len( enabled()) == 1:
+                            sn = list( enabled())[0]
+                            break
+                        time.sleep( 1 )
+                    if not sn:
+                        log.w( 'Could not recognise device in port', port)
+                    else:
+                        device = _device_by_sn.get( sn )
+                        if device:
+                            log.d( 'port', port, 'has device', sn )
+                            device._port = port
```

Comment:
We can stop if we've identified all device ports that were unknown

Reply:
if we've identified all ports we are on the last iteration anyway

Reply:
If we're missing one device port and we have two unknown, we can stop as soon as we find one device port, so we can save one acroname enable